src/vision/status_quo/alan_wants_prefetch_iterator.md

```diff
@@ -14,7 +14,51 @@ The original service consumes messages from a source stream (e.g. Kafka), proces
 
 Since messages might arrive with some delays between them, or can end at some point for a while, their number is unknown, there's an async iterator which reads the input stream and waits some time before producing a batch if the next message isn't immediately ready.
 
-While this kind of iterator returns control to the caller, and the caller continues execution, there can be a running future preparing the next batch to be taken synchronously with a very high probability.
+Alan explored `FeatureExt` from `async-std` and found no evidence that it's possible to wait for multiple features returning different results (it's not possible for `ValueTask`s in .NET, but it worked well with `Task`s which can be awaited multiple times). Later he was suggested to use an `enum` and the `race` method to achive his goal:
+
+```rust
+enum Choices<A, B, C> {
+    A(A),
+    B(B),
+    C(C),
+}
+
+// convert each future into the type `Choices<...>`:
+let future_a = async move { A(future_a.await) };
+let future_b = async move { B(future_b.await) };
+let future_c = async move { C(future_c.await) };
+
+// await the race:
+match future_a.race(future_b).race(future_c).await {
+    A(a) => ...,
+    B(b) => ....,
+    C(c) => ...,
+}
+```
+
+While that helped Alan, it was completely unobvius to him. He expected to see a macro accepting futures and producing a new future to be awaited:
+
+```rust
+match race!(feature_a, feature_b, feature_c).await {
+    // ...
+}
+```
+
+Having `join!` would be nice too for Alan, so he can avoid binding variables to features which later shall be awaited:
```

Comment:
FWIW I think [this exists](https://docs.rs/futures/0.3.15/futures/macro.join.html), but it seems like the point is that it is quite undiscoverable.

Reply:
Good, I named the macro `join` only because saw a method with the same name and meaning. I think that lack of `await` here might bring some confusion since any feature is awaited and there`s a keyword for that you type each time.